### PR TITLE
bleik/common transformers utils update

### PR DESCRIPTION
**Pull request by saidbleik**
### Description

Changed the order of moving a model to a device and constructing the optimizer.
ref: https://pytorch.org/docs/master/optim.html

### Related Issues
#512 #532 

### Checklist:

- [x] My code follows the code style of this project, as detailed in our [contribution guidelines](../CONTRIBUTING.md).
- [x] I have added tests.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by hlums** [comment]
@saidbleik Conflict needs to be resolved for tests to run.

File: utils_nlp/models/transformers/sequence_classification.py [hunk 1]
```diff
@@ -220,6 +212,9 @@ def create_dataloader_from_df(
                 max_len=max_len,
             )
 
+    def dataloader_from_dataset(
```
[the remark this answers is not in the document]
Comment (saidbleik) [reply utils_nlp/models/transformers/sequence_classification.py]: ah yes, we can move this one.

File: utils_nlp/common/pytorch_utils.py [hunk 2]
```diff
@@ -32,6 +32,46 @@ def get_device(
...
+        )
+    else:
+        if num_gpus > 1:
+            if gpu_ids is None:
+                gpu_ids = list(range(num_gpus))
+            model = torch.nn.DataParallel(model, device_ids=gpu_ids)
+    # move to device
+    return model.to(device)
+
+
```
[the remark this answers is not in the document]
Comment (saidbleik) [reply utils_nlp/common/pytorch_utils.py]: I don't think so. I can remove it.

File: utils_nlp/models/transformers/sequence_classification.py [hunk 3]
```diff
@@ -250,7 +245,10 @@ def fit(
         self,
         train_dataloader,
```
[the remark this answers is not in the document]
Comment (hlums) [reply utils_nlp/models/transformers/sequence_classification.py]: After a second thought, maybe it's hard to make this consistent for all scenarios. Let's just do what's the best for each scenario. :)

File: utils_nlp/models/transformers/common.py [hunk 4]
```diff
@@ -81,99 +81,86 @@ def set_seed(seed, cuda=True):
...
+                "weight_decay": 0.0,
+            },
+        ]
+        optimizer = AdamW(optimizer_grouped_parameters, lr=learning_rate, eps=adam_epsilon)
+        return optimizer
+
+    @staticmethod
+    def get_default_scheduler(
+        optimizer, warmup_steps, data_loader, max_steps, num_epochs, gradient_accumulation_steps
+    ):
```
Comment (daden-ms) [inline utils_nlp/models/transformers/common.py]: Can we move the logic to determine the max_steps out of get_default_scheduler? get_default_scheduler only need optimizer, max_steps, warmup_steps.

File: utils_nlp/models/transformers/common.py [hunk 5]
```diff
@@ -81,99 +81,86 @@ def set_seed(seed, cuda=True):
...
         global_step = 0
         tr_loss = 0.0
+        self.model.train()
         self.model.zero_grad()
         train_iterator = trange(
             int(num_train_epochs), desc="Epoch", disable=local_rank not in [-1, 0] or not verbose
         )
 
+        # train
         for _ in train_iterator:
```
Comment (daden-ms) [inline utils_nlp/models/transformers/common.py]: need to merge the changes from staging

File: utils_nlp/models/transformers/common.py (hunk 4 above)
Comment (daden-ms) [reply utils_nlp/models/transformers/common.py]: we can wrap the logic of determining the max_steps as a function too.

Comment (saidbleik) [reply utils_nlp/models/transformers/common.py]: Yes, I actually did that last week. Let me push those.

File: utils_nlp/models/transformers/common.py [hunk 6]
```diff
@@ -176,39 +149,47 @@ def fine_tune(
...
 
         return global_step, tr_loss / global_step
 
     def predict(self, eval_dataloader, device, get_inputs, verbose=True):
         self.model.eval()
         for batch in tqdm(eval_dataloader, desc="Evaluating", disable=not verbose):
-            batch = tuple(t.to(device) for t in batch)
             with torch.no_grad():
-                inputs = get_inputs(batch, self.model_name, train_mode=False)
+                inputs = get_inputs(batch, device, self.model_name, train_mode=False)
```
Comment (daden-ms) [inline utils_nlp/models/transformers/common.py]: In the latest staging branch, there is move_batch_to_device in the fine-tune and predict functions. I am not sure why it's not showing us in the staging branch of your pull request.

Comment (saidbleik) [reply utils_nlp/models/transformers/common.py]: I removed this function and included the logic in get_inputs()

Comment (daden-ms) [reply utils_nlp/models/transformers/common.py]: I looked into your branch, the summarization work is not included: https://github.com/microsoft/nlp-recipes/tree/bleik/optim-patch/utils_nlp/models/transformers  Is this the place I should look at?

Comment (saidbleik) [reply utils_nlp/models/transformers/common.py]: right, I haven't included it yet. It needs to be modified to match the changes.

File: utils_nlp/models/transformers/sequence_classification.py [hunk 7]
```diff
@@ -307,20 +249,44 @@ def fit(
             seed (int, optional): Random seed used to improve reproducibility. Defaults to None.
         """
 
+        # get device
+        device, num_gpus = get_device(num_gpus=num_gpus, local_rank=local_rank)
+
+        # move model
+        self.model = move_model_to_device(self.model, device, num_gpus, gpu_ids, local_rank)
```
Comment (hlums) [inline utils_nlp/models/transformers/sequence_classification.py]: According to https://github.com/huggingface/transformers/blob/715fa638a7601eef9fd0ae9311542e1adc56ef8d/examples/run_glue.py#L152, the multi-gpu training code should be after the amp initialization. Right now, this is done in the child class before calling the `fine_tune` method where the amp initialization happens.

File: utils_nlp/models/transformers/common.py [hunk 8]
```diff
@@ -223,15 +165,10 @@ def move_batch_to_device(batch, device):
                 if (step + 1) % gradient_accumulation_steps == 0:
```
Comment (hlums) [inline utils_nlp/models/transformers/common.py]: Also maybe the gradient clipping should be done after accumulating gradients like https://github.com/huggingface/transformers/blob/dfe012ad9d6b6f0c9d30bc508b9f1e4c42280c07/examples/run_glue.py#L228

File: utils_nlp/models/bert/sequence_classification.py [hunk 9]
```diff
@@ -13,7 +13,7 @@
 from tqdm import tqdm
 
 from utils_nlp.models.bert.common import Language
```
Comment (daden-ms) [inline utils_nlp/models/bert/sequence_classification.py]: Do we still want to keep the models/bert folder? I thought it's going to be replaced by the transformer folder.

File: examples/text_classification/tc_mnli_transformers.ipynb [hunk 10]
```diff
@@ -518,11 +604,17 @@
...
     "        cache_dir=CACHE_DIR,\n",
     "    )\n",
-    "    train_dataloader = processor.create_dataloader_from_df(\n",
-    "        df_train, TEXT_COL, LABEL_COL, max_len=MAX_LEN, batch_size=BATCH_SIZE, num_gpus=NUM_GPUS, shuffle=True\n",
+    "    train_dataset = processor.dataset_from_dataframe(\n",
+    "        df_train, TEXT_COL, LABEL_COL, max_len=MAX_LEN\n",
     "    )\n",
-    "    test_dataloader = processor.create_dataloader_from_df(\n",
-    "        df_test, TEXT_COL, LABEL_COL, max_len=MAX_LEN, batch_size=BATCH_SIZE, num_gpus=NUM_GPUS, shuffle=False\n",
+    "    train_dataloader = dataloader_from_dataset(\n",
```
Comment (daden-ms) [inline examples/text_classification/tc_mnli_transformers.ipynb]: can we decide what data loader is needed in the fit and predict function so that we don't need to repeat the logic of handling distributed training in both data loader function and the fit function? We can just pass in the dataset for both fit and predict function.

File: utils_nlp/models/bert/sequence_classification.py (hunk 9 above)
Comment (saidbleik) [reply utils_nlp/models/bert/sequence_classification.py]: I'll remove it if no one needs it - let's double check first

File: utils_nlp/common/pytorch_utils.py [hunk 11]
```diff
@@ -32,59 +28,107 @@ def get_device(
...
+    # move to device
+    return model.to(device)
+
 
+def dataloader_from_dataset(ds, batch_size=32, num_gpus=None, shuffle=False, distributed=False):
+    """Creates a PyTorch DataLoader given a Dataset object.
+
+    Args:
+        ds (torch.utils.data.DataSet): A PyTorch dataset.
+        batch_size (int, optional): Batch size. Defaults to 32.
```
Comment (hlums) [inline utils_nlp/common/pytorch_utils.py]: Can explain batch_size is batch_size per gpu in the docstring? Would be even better to name the argument batch_size_per_gpu to make it more explicit, but I'll leave it up to you.

File: utils_nlp/models/transformers/named_entity_recognition.py [hunk 12]
```diff
@@ -42,27 +40,36 @@ def __init__(self, model_name="bert-base-cased", to_lower=False, cache_dir="."):
...
+            train_mode (bool, optional): Training mode flag.
+                Defaults to True.
 
         Returns:
-            dict: A dictionary object contains all needed information for training or testing.
+            dict: Dictionary containing input ids, segment ids, masks, and labels.
+                Labels are only returned when train_mode is True.
         """
+        batch = tuple(t.to(device) for t in batch)
+        if model_name.split("-")[0] in ["bert", "distilbert"]:
```
[the remark this answers is not in the document]
Comment (saidbleik) [reply utils_nlp/models/transformers/named_entity_recognition.py]: these are model specific checks - can't group them all under supported models list (see line 67)

File: utils_nlp/models/transformers/extractive_summarization.py [hunk 13]
```diff
@@ -354,6 +348,14 @@ def get_inputs(batch, model_name, train_mode=True):
                     "labels": batch.labels,
                 }
             else:
+                batch["src"] = batch["src"].to(device)
```
[the remark this answers is not in the document]
Comment (saidbleik) [reply utils_nlp/models/transformers/extractive_summarization.py]: perhaps we can remove the Bunch wrapper @daden-ms ?

File: utils_nlp/models/transformers/common.py [hunk 14]
```diff
@@ -246,31 +194,20 @@ def move_batch_to_device(batch, device):
                     epoch_iterator.close()
                     break
 
-            # empty cache
-            torch.cuda.empty_cache()
```
[the remark this answers is not in the document]
Comment (saidbleik) [reply utils_nlp/models/transformers/common.py]: It's not recommended. Cache should be cleared automatically.

File: utils_nlp/models/transformers/extractive_summarization.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (daden-ms) [reply utils_nlp/models/transformers/extractive_summarization.py]: "batch = batch.to(device)" should work and there is no need to move each field one by one. It's not related to Bunch.

[the remark this answers is not in the document]
Comment (saidbleik) [reply utils_nlp/models/transformers/extractive_summarization.py]: I cleared these lines. This was copied as is from the previous version.

File: utils_nlp/models/transformers/named_entity_recognition.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (hlums) [reply utils_nlp/models/transformers/named_entity_recognition.py]: Yes. Line 67 is model specific check, but the else statement after this if means we are checking if the model_name is supported by this class.
